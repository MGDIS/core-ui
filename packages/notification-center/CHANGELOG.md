# @mgdis/notification-center

## 1.1.13

### Patch Changes

- c567b43: update dependencies

## 1.1.12

### Patch Changes

- 43f1421: update dependancies
- 28782b6: [#487](https://gitlab.mgdis.fr/core/core-ui/core-ui/-/issues/487) notification-center: update z-index to improve bootstrap compatibility

## 1.1.11

### Patch Changes

- 253c7a6: Update dependencies

## 1.1.10

### Patch Changes

- d674c81: Update dependencies

## 1.1.9

### Patch Changes

- a8d2eea: Update dependencies

## 1.1.8

### Patch Changes

- fbdf99a: Update dependencies

## 1.1.7

### Patch Changes

- 7a7e0ec: Update dependencies

## 1.1.6

### Patch Changes

- c87d711: Update dependencies

## 1.1.5

### Patch Changes

- eb3b1ab: Update dependencies

## 1.1.4

### Patch Changes

- bad9447: [#450](https://gitlab.mgdis.fr/core/core-ui/core-ui/-/issues/450) notification-center: Notification center not displaying on angularJS app

## 1.1.3

### Patch Changes

- 558821cc: add Figma import script

## 1.1.2

### Patch Changes

- c25283d: [#298](https://gitlab.mgdis.fr/core/core-ui/core-ui/-/issues/298) chore: create apps to test lib integration

## 1.1.1

### Patch Changes

- 1b13754: [#344](https://gitlab.mgdis.fr/core/core-ui/core-ui/-/issues/344) mg-icon - SVG are not displayed on safari

## 1.1.0

### Minor Changes

- 067160b: Add UMD export

### Patch Changes

- cd9dfb6: change exposed global variable name for umd build

## 1.0.1

### Patch Changes

- 7ff4d3c: Update screenshots

## 1.0.0

### Major Changes

- 6fe6663: notification center release
