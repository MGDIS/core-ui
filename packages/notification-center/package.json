{
  "name": "@mgdis/notification-center",
  "version": "1.1.4",
  "description": "Display notifications throught iframes",
  "author": "Simon Duhem <duhem-s@mgdis.fr>",
  "license": "BSD-3-Clause",
  "types": "dist/src/index.d.ts",
  "scripts": {
    "build": "tsc --noEmit && vite build",
    "test:unit": "jest --coverage",
    "test:e2e": "pnpm --package=@pnpm/@playwright/test dlx playwright test",
    "test:e2e:docker": "pnpm dlx @playwright/test test",
    "test:watch": "jest --watch",
    "coverage": "jest --coverage",
    "lint": "pnpm lint:es && pnpm lint:prettier",
    "lint:fix": "pnpm lint:es:fix && pnpm lint:prettier:fix",
    "lint:es": "eslint 'src/**/*.{ts,tsx}' --ignore-path ../../.gitignore",
    "lint:es:fix": "eslint 'src/**/*.{ts,tsx}' --fix --ignore-path ../../.gitignore",
    "lint:prettier": "prettier --check '**/*.{js,ts,tsx,scss,html,mdx,json}' --ignore-path ../../.gitignore",
    "lint:prettier:fix": "prettier --write '**/*.{js,ts,tsx,scss,html,mdx,json}' --ignore-path ../../.gitignore"
  },
  "files": [
    "dist"
  ],
  "main": "./dist/notification-center.umd.js",
  "module": "./dist/notification-center.es.js",
  "exports": {
    ".": {
      "types": "./dist/notification-center.d.ts",
      "import": "./dist/notification-center.es.js",
      "require": "./dist/notification-center.umd.js"
    }
  },
  "devDependencies": {
    "@mgdis/mg-components": "workspace:*",
    "@mgdis/playwright-helpers": "workspace:*",
    "@playwright/test": "1.41.1",
    "@types/jest": "29.5.11",
    "@types/sanitize-html": "2.9.5",
    "eslint-config-custom": "workspace:*",
    "jest": "29.7.0",
    "jest-config": "workspace:*",
    "jest-environment-jsdom": "29.7.0",
    "jest-serializer-html": "7.1.0",
    "prettier": "3.2.4",
    "prettier-config": "workspace:*",
    "ts-jest": "29.1.2",
    "tsconfig": "workspace:*",
    "typescript": "5.3.3",
    "vite": "5.0.12",
    "vite-plugin-dts": "3.7.2"
  },
  "dependencies": {
    "sanitize-html": "2.11.0"
  }
}
