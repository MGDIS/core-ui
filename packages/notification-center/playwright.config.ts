import base from '@mgdis/playwright-config';
import type { PlaywrightTestConfig } from '@playwright/test';

const config: PlaywrightTestConfig = {
  ...base,
  /* Maximum time one test can run for. */
  timeout: 60 * 1000,
  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'cd ../.. && pnpm apps:notification-center',
    port: 3210,
    reuseExistingServer: true,
  },
};

export default config;
