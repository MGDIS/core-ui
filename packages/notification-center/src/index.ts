export { NotificationCenter } from './notification-center';
