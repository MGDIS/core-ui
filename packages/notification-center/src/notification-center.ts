/* eslint-disable no-console */
import type { NotificationData } from './types';
import sanitizeHtml from 'sanitize-html';

/**
 * NotificationCenter class
 */
class NotificationCenter {
  #rootWindow: Window;

  #notificationsReceiver: HTMLDivElement = document.createElement('div');

  readonly #appId: string = 'mg-notification-center';

  constructor() {
    // Get Root Window
    this.#rootWindow = this.#getRootWindow(window);
    // If the window is not in an iframe
    if (!this.#isInIframe(window)) {
      // Check if mg-components is loaded
      if (customElements.get('mg-alert') === undefined) {
        console.error('mg-components is not loaded.');
      }
      // We listen to events
      this.#addEventListener(this.#rootWindow, ({ data }: { data: NotificationData }) => {
        if (data.appId === this.#appId) this.#displayNotification(data);
      });
      // When DOM is ready we add our CSS and the notifications zone
      if (this.#rootWindow.document.readyState === 'complete') this.#render();
      this.#rootWindow.addEventListener('DOMContentLoaded', () => {
        this.#render();
      });
    }
  }

  /**
   * Render notification-center in rootwindow document
   */
  #render = () => {
    // If notification center is not already created
    if (this.#rootWindow.document.getElementById(this.#appId) === null) {
      // Add style
      const css: Text = document.createTextNode(
        `#${this.#appId} {
          position: fixed;
          bottom: 0;
          right: 0;
          padding:  0 1rem 1rem;
          width: 40rem;
          max-width: 100%;
          display: flex;
          flex-direction: column;
          align-items: flex-end;
          gap: 0.5rem;
          pointer-events: none;
          z-index: 1000;
        }
        #${this.#appId} mg-alert {
          pointer-events: auto;
          animation: append-item .4s ease;
        }
        @media (prefers-reduced-motion) {
          #${this.#appId} mg-alert {
            animation: none;
          }
        }
        @keyframes append-item {
          0% {
            opacity: 0;
            transform: translateY(50%);
          }
          100% {
            opacity: 1;
            transform: translateY(0);
          }
        }`,
      );
      const style: HTMLStyleElement = document.createElement('style');
      style.appendChild(css);
      this.#rootWindow.document.head.appendChild(style);
      // Add div to receive notifications
      this.#notificationsReceiver.innerHTML = '';
      this.#notificationsReceiver.id = this.#appId;
      this.#rootWindow.document.body.appendChild(this.#notificationsReceiver);
    }
  };

  /**
   * Check if the window is in an iframe
   *
   * @param localWindow - the window that load the script
   * @returns the window is in an iframe
   */
  #isInIframe = (localWindow: Window): boolean => localWindow.self !== localWindow.top;

  /**
   * Add event listener on window
   *
   * @param targetWindow - window that will listen to events
   * @param callback - method to execute when we receive an event
   */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  #addEventListener = (targetWindow: Window, callback: any): void => {
    targetWindow.addEventListener('message', callback);
  };

  /**
   * Post notification to root window
   *
   * @param notification - notification data
   */
  postMessage = (notification: NotificationData): void => {
    this.#rootWindow.postMessage({ ...notification, appId: this.#appId }, document.location.origin);
  };

  /**
   * Get the root window
   *
   * @param localWindow - the window where we want to identify the root
   * @returns the root window
   */
  #getRootWindow = (localWindow: Window): Window => {
    // Check if we have permission to access parent
    try {
      const parentDocument = localWindow.parent.document;
      if (!parentDocument) throw new Error('Cannot access parent document');
    } catch (err) {
      console.error('Different hosts between iframes:', err);
      return localWindow;
    }
    return this.#isInIframe(localWindow) ? this.#getRootWindow(localWindow.parent) : localWindow;
  };

  /**
   * Display notification
   *
   * @param notification - notification data
   */
  #displayNotification = ({ content, variant, delay, context }: NotificationData): void => {
    // Remove mg-alert with same context
    if (context)
      this.#notificationsReceiver.querySelectorAll(`mg-alert[data-mg-alert-context='${context}']`).forEach(mgAlert => {
        mgAlert.remove();
      });
    // Init mg-alert
    const mgAlert: HTMLElement = document.createElement('mg-alert');
    mgAlert.setAttribute('close-button', '');
    // Variant
    if (variant) mgAlert.setAttribute('variant', variant);
    // Delay
    if (delay === undefined && variant === 'success') delay = 5;
    else if (delay !== undefined && delay < 3) delay = undefined;
    if (delay) mgAlert.setAttribute('delay', delay.toString());
    // Context
    if (context) mgAlert.dataset.mgAlertContext = context;
    // Remove notification on close
    mgAlert.addEventListener('component-hide', () => {
      mgAlert.remove();
    });
    // Add content
    mgAlert.innerHTML = sanitizeHtml(content);
    // Add mg-alert
    this.#notificationsReceiver.appendChild(mgAlert);
  };
}

export { NotificationCenter };
