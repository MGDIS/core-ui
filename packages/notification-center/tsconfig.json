{
  "extends": "tsconfig/base.node18.json",
  "compilerOptions": {
    "lib": ["ES2015", "dom"],
    "outDir": "dist"
  },
  "include": ["src/**/*.ts"],
  "exclude": ["src/**/*.spec.ts", "src/**/*.e2e.ts"]
}
