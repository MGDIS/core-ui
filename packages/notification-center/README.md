# Notification center

Notification center is the new way to display notifications in MGDIS products.

Considering notifications as :

- an information message
- an error message
- a warning message
- a success message

## Why?

Because of the iframes we had to display notifications in the window and close by the origin action was. In some case the notification is not directly visible, we have to scroll to see it for example.

We had to add `mg-message` in almost every form.

It has some concistency issues.

## How?

The idea is to load the library on every frontend projects, if it's loaded in the top window it will listen to messages, so even if you are running your project localy, the notification will be displayed.

The library display notifications using [mg-message](http://core.pages.mgdis.fr/core-ui/core-ui/?path=/story/molecules-mg-message--mg-message) from [mg-components](http://core.pages.mgdis.fr/core-ui/core-ui/).

### Include the lib

If you just need to listen to messages to display notifications:

```TS
import { NotificationCenter } from '@mgdis/notification-center';
new NotificationCenter();
```

To send messages :

```TS
import { NotificationCenter } from '@mgdis/notification-center';
const notif = new NotificationCenter();

// Default (minimum) message
notif.postMessage({
  content: 'Message notification'
});
```

Here is the full arguments list:

- **content:** The message you want to display. HTML content will be sanitized.
- **variant:** Set the message variant type.  
  Default `'info'`, can also be `'danger'`, `'success'`, `'warning'`.
- **delay:** Define the number of second the message will be displayed.  
   Default values correspond to the expected behavior so these values should only be changed if it has been explicitly requested.

  - For `'info'`, `'danger'` and `'warning'` variants, default value is `undefined`.
  - For `'success'` variant default value is `5`. Can be set to `undefined` by setting its value to `0`.

  Minimum value is `3`.

- **context:** Previous notification using the same context value will be removed.  
  Default: `undefined`

```TS
// Possibilities
notif.postMessage({
  content: 'Message notification',
  variant: 'danger',
  delay: 10
  context: 'form-context-id',
});
```

## Possible improvements

- Add mg-details in message.
- Add button or link in message.
- Loading message.
- Add possibility to define message zone:
  - top-left, top, top-right, bottom-right, bottom, bottom-left
  - x,y spacing
- Add confirmation modale.
