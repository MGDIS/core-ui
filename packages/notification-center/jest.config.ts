import base from 'jest-config/base';
import { name } from './package.json';
import type { Config } from 'jest';

const config: Config = {
  ...base,
  preset: 'ts-jest',
  testEnvironment: 'jsdom',
  snapshotSerializers: ['jest-serializer-html'],
  displayName: name,
};

export default config;
