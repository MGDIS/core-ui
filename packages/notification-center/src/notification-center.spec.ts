import { defineCustomElements } from '@mgdis/mg-components/loader';
import { NotificationCenter } from './notification-center';

describe('Notification center', () => {
  let notifsCenter;

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    // Clean jsdom
    document.body.innerHTML = '';
    document.head.innerHTML = '';
  });

  it('Should log an error if custom element are not loaded', () => {
    const spyConsole = jest.spyOn(console, 'error');

    notifsCenter = new NotificationCenter();

    expect(notifsCenter).toBeDefined();
    expect(spyConsole).toHaveBeenCalledWith('mg-components is not loaded.');
  });

  it('Should throw an error when cannot access parent', () => {
    const spyConsole = jest.spyOn(console, 'error');
    const spyWindowParent = jest.spyOn(window, 'parent', 'get').mockImplementationOnce(() => {
      throw new Error('non');
    });

    notifsCenter = new NotificationCenter();

    expect(notifsCenter).toBeDefined();
    expect(spyWindowParent).toHaveBeenCalled();
    expect(spyConsole.mock.calls[0]?.[0]).toEqual('Different hosts between iframes:');
  });

  it('Should throw an error when parent document is undefined', () => {
    const spyConsole = jest.spyOn(console, 'error');
    const spyWindowParent = jest.spyOn(window, 'parent', 'get').mockImplementationOnce(() => ({} as Window));

    notifsCenter = new NotificationCenter();

    expect(notifsCenter).toBeDefined();
    expect(spyWindowParent).toHaveBeenCalled();
    expect(spyConsole.mock.calls[0]?.[0]).toEqual('Different hosts between iframes:');
  });

  describe('defined custom elements', () => {
    beforeAll(() => {
      // load mg-components
      defineCustomElements(window);
    });

    beforeEach(() => {
      notifsCenter = new NotificationCenter();
      expect(notifsCenter).toBeDefined();
    });

    it('Should init the lib', () => {
      // Trigger DOM ready event
      window.dispatchEvent(new Event('DOMContentLoaded'));
      expect(document.head.innerHTML).toContain('<style>#mg-notification-center');
      expect(document.getElementById('mg-notification-center')).not.toBeNull();
    });

    it.each([
      { content: 'Default example' },
      { content: 'Danger variant', variant: 'danger' },
      { content: 'Warning variant', variant: 'warning' },
      { content: 'Success variant', variant: 'success' },
      { content: 'With delay', delay: 5 },
      { content: 'With context', context: 'blu' },
    ])('Should display a notification %s', eventData => {
      window.dispatchEvent(new Event('DOMContentLoaded'));
      window.dispatchEvent(
        new MessageEvent('message', {
          data: { ...eventData, appId: 'mg-notification-center' },
        }),
      );
      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('Should remove unsafe or unknown html', () => {
      window.dispatchEvent(new Event('DOMContentLoaded'));
      window.dispatchEvent(
        new MessageEvent('message', {
          data: {
            content: `<p>Content with unsafe or unknown html</p>
<script>alert("blu")<script>
<not-tag>not a tag</not-tag>`,
            appId: 'mg-notification-center',
          },
        }),
      );
      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('Should post a message', () => {
      const messageData = { content: 'Default example' };
      const spyPostMessage = jest.spyOn(window, 'postMessage');
      notifsCenter.postMessage(messageData);
      expect(spyPostMessage).toHaveBeenCalledWith({ ...messageData, appId: 'mg-notification-center' }, '*');
    });

    it('Should remove previous message with same context', () => {
      window.dispatchEvent(new Event('DOMContentLoaded'));
      window.dispatchEvent(
        new MessageEvent('message', {
          data: {
            content: 'First message with context',
            context: 'blu',
            appId: 'mg-notification-center',
          },
        }),
      );
      expect(document.body.innerHTML).toMatchSnapshot();
      window.dispatchEvent(
        new MessageEvent('message', {
          data: {
            content: 'Second message with same context',
            context: 'blu',
            appId: 'mg-notification-center',
          },
        }),
      );
      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('Should remove message on close', () => {
      window.dispatchEvent(new Event('DOMContentLoaded'));
      window.dispatchEvent(
        new MessageEvent('message', {
          data: {
            content: 'Message to close',
            appId: 'mg-notification-center',
          },
        }),
      );
      expect(document.body.innerHTML).toMatchSnapshot();
      const mgMessage = document.querySelector('mg-message');
      mgMessage?.dispatchEvent(new CustomEvent('component-hide', { bubbles: true }));
      expect(document.body.innerHTML).toMatchSnapshot();
    });
  });

  it('Should fake iframe', () => {
    const spyWindowSelf = jest.spyOn(window, 'self', 'get').mockImplementationOnce(jest.fn());
    const spyWindowParent = jest.spyOn(window, 'parent', 'get').mockImplementation(() => window);

    notifsCenter = new NotificationCenter();

    expect(notifsCenter).toBeDefined();
    // Check 3 times if it's an iframe
    expect(spyWindowSelf).toHaveBeenCalledTimes(3);
    expect(spyWindowParent).toHaveBeenCalledTimes(3);
  });
});
