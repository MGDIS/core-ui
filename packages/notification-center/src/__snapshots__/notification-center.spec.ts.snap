// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification center defined custom elements Should display a notification {
  content: 'Success variant without delay',
  variant: 'success',
  delay: 0
} 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            variant="success"
  >
    Success variant without delay
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'Danger variant', variant: 'danger' } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            variant="danger"
  >
    Danger variant
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'Default example' } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button>
    Default example
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'Success variant', variant: 'success' } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            variant="success"
            delay="5"
  >
    Success variant
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'Warning variant', variant: 'warning' } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            variant="warning"
  >
    Warning variant
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'With context', context: 'blu' } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            data-mg-alert-context="blu"
  >
    With context
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should display a notification { content: 'With delay', delay: 5 } 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            delay="5"
  >
    With delay
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should remove message on close 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button>
    Message to close
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should remove message on close 2`] = `
<div id="mg-notification-center">
</div>
`;

exports[`Notification center defined custom elements Should remove previous message with same context 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            data-mg-alert-context="blu"
  >
    First message with context
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should remove previous message with same context 2`] = `
<div id="mg-notification-center">
  <mg-alert close-button
            data-mg-alert-context="blu"
  >
    Second message with same context
  </mg-alert>
</div>
`;

exports[`Notification center defined custom elements Should remove unsafe or unknown html 1`] = `
<div id="mg-notification-center">
  <mg-alert close-button>
    <p>
      Content with unsafe or unknown html
    </p>
  </mg-alert>
</div>
`;
